Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,459 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+// This is a companion to System.swift that provides only Linux specials: either things that exist
+// only on Linux, or things that have Linux-specific extensions.
+
+#if os(Linux)
+
+import CNIOLinux
+
+// we stuff the event type into the user data for the sqe together with
+// the fd to match the events without needing any memory allocations or
+// references. Just shift in the event type in the upper 32 bits.
+
+internal enum CqeEventType : Int {
+    case poll = 1, pollModify, pollDelete // start with 1 to not get zero bit patterns for stdin
+}
+
+internal enum UringError: Error {
+    case loadFailure
+    case uringSetupFailure
+    case uringWaitCqeFailure
+    case uringWaitCqeTimeoutFailure
+}
+
+internal extension TimeAmount {
+    func kernelTimespec() -> __kernel_timespec {
+        var ts = __kernel_timespec()
+        ts.tv_sec = self.nanoseconds / 1_000_000_000
+        ts.tv_nsec = self.nanoseconds % 1_000_000_000
+        return ts
+    }
+}
+
+public struct UringEvent {
+    var fd : Int32
+    var pollMask : UInt32
+}
+
+public class Uring {
```

Comment:
Presumably little of this wants to be `public`.

Reply:
Agree, still some left, will fix.